README.md
# permission

[![Build Status](https://travis-ci.org/lykmapipo/permission.svg?branch=master)](https://travis-ci.org/lykmapipo/permission)
[![Dependencies Status](https://david-dm.org/lykmapipo/permission/status.svg?style=flat-square)](https://david-dm.org/lykmapipo/permission)
[![Deploy to Heroku](https://www.herokucdn.com/deploy/button.png)](https://heroku.com/deploy?template=https://github.com/lykmapipo/permission/tree/master)

A representation of ability to perform an action or access a resource.

Permissions are immutable and reflect an application's raw functionality (opening files, printing files, creating users, etc). This is what allows a system's security policy to be dynamic: because Permissions represent raw functionality and only change when the application's source code changes, they are immutable at runtime - they represent 'what' the system can do.

[Demo](https://permision.herokuapp.com/v1/permissions)

## Requirements

- [nodejs v8.11.1+](https://nodejs.org)
- [npm v5.6.0+](https://www.npmjs.com/)
- [MongoDB v3.4.10+](https://www.mongodb.com/)
- [mongoose v5.2.5+](https://github.com/Automattic/mongoose)

## Installation

```sh
npm install @lykmapipo/permission --save
```

## Usage

```js
const { connect } = require('@lykmapipo/mongoose-common');
const { start, mount } = require('@lykmapipo/express-common');
const { Permission, permissionRouter } = require('@lykmapipo/permission');

// connect to mongodb
connect(process.env.MONGODB_URI, error => { ... });

// mount permission http router
mount(permissionRouter);

// fire the http server
start(error => { ... });
```

## Demo
[View Demo](https://permision.herokuapp.com/v1/permissions)
[View apidoc](https://lykmapipo.github.io/permission/)

## Testing

- Clone this repository

- Install all development dependencies

```sh
npm install
```

- Run example

```sh
npm run dev
```

- Then run test

```sh
npm test
```

## Contribute

It will be nice, if you open an issue first so that we can know what is going on, then, fork this repo and push in your ideas. Do not forget to add a bit of test(s) of what value you adding.

## References
- [Role Based Access Control](https://en.wikipedia.org/wiki/Role-based_access_control)
- [Apache Shiro - Permission](https://shiro.apache.org/permissions.html)

## Licence

The MIT License (MIT)

Copyright (c) lykmapipo & Contributors

Permission is hereby granted, free of charge, to any person obtaining a copy of this software and associated documentation files (the “Software”), to deal in the Software without restriction, including without limitation the rights to use, copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the Software, and to permit persons to whom the Software is furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED “AS IS”, WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.


CHANGELOG.md
#### 0.8.35 (2020-12-21)

#### 0.8.34 (2020-09-16)

#### 0.8.33 (2020-07-30)

#### 0.8.32 (2020-07-14)

#### 0.8.31 (2020-07-02)

#### 0.8.30 (2020-06-23)

##### Chores

* **deps:**  force latest version & audit fix ([18f58212](https://github.com/lykmapipo/permission/commit/18f582120c098aaeb34e527cd1580c2a60bce4f8))

##### Documentation Changes

* **changelog:**  generate latest release notes ([59ab23c0](https://github.com/lykmapipo/permission/commit/59ab23c0a3230abb8dc4fa101e42fa17643c401e))

#### 0.8.29 (2020-06-11)

#### 0.8.28 (2020-05-20)

#### 0.8.27 (2020-05-10)

#### 0.8.26 (2020-05-03)

#### 0.8.25 (2020-05-01)

#### 0.8.24 (2020-04-27)

#### 0.8.23 (2020-04-18)

#### 0.8.22 (2020-03-30)

#### 0.8.21 (2020-03-06)

#### 0.8.20 (2020-02-28)

#### 0.8.19 (2020-02-16)

#### 0.8.18 (2020-02-04)

#### 0.8.17 (2020-01-22)

#### 0.8.16 (2020-01-17)

#### 0.8.15 (2020-01-12)

#### 0.8.14 (2019-12-17)

#### 0.8.13 (2019-12-13)

#### 0.8.12 (2019-10-23)

#### 0.8.11 (2019-10-10)

#### 0.8.10 (2019-09-28)

#### 0.8.9 (2019-09-26)

#### 0.8.8 (2019-09-16)

#### 0.8.7 (2019-08-20)

#### 0.8.6 (2019-08-15)

#### 0.8.5 (2019-07-09)

#### 0.8.4 (2019-06-19)

#### 0.8.3 (2019-06-10)

#### 0.8.2 (2019-06-08)

#### 0.8.1 (2019-05-20)

#### 0.8.0 (2019-05-19)

##### Chores

* **deps:**
  *  add express-rest-actions & audit fix ([502930b2](https://github.com/lykmapipo/permission/commit/502930b2dbc9741697a7a9e019613a7251673127))
  *  remove unused dependencies ([7ae7fa4f](https://github.com/lykmapipo/permission/commit/7ae7fa4f7d9987290b9a4975eaf29af4c2b8fdcb))
  *  force latest versions & audit fix ([092d3631](https://github.com/lykmapipo/permission/commit/092d3631025a6a3c71da51e0275da2ad454c8dca))
* **ci:**  force latest nodejs version ([d253d697](https://github.com/lykmapipo/permission/commit/d253d697ee216afc33efcc710bc1b22607ecd97e))
*  upgrade examples ([c95607b6](https://github.com/lykmapipo/permission/commit/c95607b60debdf9941b0743600e80a0e700996d1))

##### Documentation Changes

*  remove year from license ([c9637b22](https://github.com/lykmapipo/permission/commit/c9637b22e78556f68cd07aec6d1cf6c23033d254))
*  update usage docs ([732ca1d6](https://github.com/lykmapipo/permission/commit/732ca1d653f1c3d7f75a6c76c97830de694f8bec))

##### New Features

*  implement resources permissions prepare ([c15cb581](https://github.com/lykmapipo/permission/commit/c15cb581be00d9a12cef03968bf3690d473326d7))
*  implement prepareSeedCriteria ([315af321](https://github.com/lykmapipo/permission/commit/315af3215e55eba93e1b3a34d12700ff97be86cc))

##### Bug Fixes

*  ensure permission package info exposed ([eb31450f](https://github.com/lykmapipo/permission/commit/eb31450f7c209ef8437c7c64a772147a87226c26))

##### Refactors

* **router:**  migrate to use express-rest-actions ([ae78530b](https://github.com/lykmapipo/permission/commit/ae78530b2629b4fa13cf4b842fded2403f33f0dd))
* **see:**  use mongoose-common seed logics ([e6a26e1a](https://github.com/lykmapipo/permission/commit/e6a26e1a005ed1fa76199945dfab0b91439cc070))
*  start upgrade to latest express-common ([76443c1e](https://github.com/lykmapipo/permission/commit/76443c1e26a1b13df433d92974b2ce22ea98541f))

##### Tests

*  fix seeding specs ([a6d2bdd7](https://github.com/lykmapipo/permission/commit/a6d2bdd7f16af4c5761f2e82d56459ce82a3e829))
*  refactor to use mongoose-test-helpers ([4b1b24d2](https://github.com/lykmapipo/permission/commit/4b1b24d2e75ca9297a6c61b3bcb2ee08343fe179))
* **integration:**
  *  refactor api specs to use express-test-helpers ([c489fbf0](https://github.com/lykmapipo/permission/commit/c489fbf0f9acdc219fa9bc61d90f0dbd5fbc342d))
  *  fix put assertions ([31d46cac](https://github.com/lykmapipo/permission/commit/31d46caceb2ba956353d89c8d2ef05513f866a30))
  *  fix patch not exist assertions ([87b6a2d3](https://github.com/lykmapipo/permission/commit/87b6a2d3178ab79d22fafadfecef51f62b4ebcda))
  *  fix patch assertions ([89e97e89](https://github.com/lykmapipo/permission/commit/89e97e89965a08af52bf3deee729b95de945b89c))
  *  fix getById assertions ([ce590a25](https://github.com/lykmapipo/permission/commit/ce590a251c61043925bc74159daa6e85610009ff))
  *  fix delete assertions ([6ca781ce](https://github.com/lykmapipo/permission/commit/6ca781ce90524a507acc1f5f252d53439b2aee08))

#### 0.7.2 (2019-04-16)

##### Chores

*  force latest dependencies ([60cbcd1c](https://github.com/lykmapipo/permission/commit/60cbcd1ce12d386a605793945b1b6b126846d932))

##### Documentation Changes

*  add code of conduct & contributing guide ([90615919](https://github.com/lykmapipo/permission/commit/90615919754aefa39a62c66de8cbf19d116267d1))

#### 0.7.1 (2019-03-28)

##### Chores

*  force latest dependencies ([e591aad5](https://github.com/lykmapipo/permission/commit/e591aad57c5d5b9229e15d5d2db79bba688daae3))

#### 0.7.0 (2019-02-14)

##### Refactors

*  return all permissions on http get ([dedbd25b](https://github.com/lykmapipo/permission/commit/dedbd25b7cd75b8101018eda7afbcabfee1eb2f1))
*  extract seed on example ([77ddd4af](https://github.com/lykmapipo/permission/commit/77ddd4af19acee047b447be97843ac68dd0f8ee4))

#### 0.6.1 (2019-01-31)

##### Refactors

*  seed stages ([1b3f16bf](https://github.com/lykmapipo/permission/commit/1b3f16bf22a5539b2013ae92eeb95187c0668dcc))
*  use latest mongoose-common ([0b5feb78](https://github.com/lykmapipo/permission/commit/0b5feb78ff17da553f51c4ecc00176ba1df560aa))




lib/permission.http.router.js
'use strict';

/**
 * @apiDefine Permission Permission
 *
 * @apiDescription A representation of ability to perform an action or
 * access a resource.
 *
 * It is important to understand a Permission instance only represents
 * functionality or access - it does not grant it.
 *
 * A Permission is the most granular, or atomic, unit in a system's security
 * policy and is the cornerstone upon which fine-grained security models
 * are built.
 *
 * Permissions are immutable and reflect an application's raw functionality
 * (opening files, printing files, creating users, etc). This is what
 * allows a system's security policy to be dynamic: because Permissions
 * represent raw functionality and only change when the application's source
 * code changes, they are immutable at runtime - they represent 'what'
 * the system can do.
 *
 * Granting access to an application functionality or a particular resource
 * is done by the application's security configuration, typically by
 * assigning Permissions to users, roles and/or groups.
 *
 * Most applications do associate a named role with permissions
 * (i.e. a role 'has a' collection of Permissions) and then associate
 * users with roles (i.e. a user 'has a' collection of roles) so that by
 * transitive association, the user 'has' the permissions in their roles.
 *
 * There are numerous variations on this theme (permissions assigned directly
 * to users, or assigned to groups, and users added to groups and these groups
 * in turn have roles, etc, etc).
 *
 * By employing a permission-based security model instead of a role-based one,
 * users, roles, and groups can all be created, configured and/or deleted at
 * runtime. This enables an extremely powerful security model.
 *
 *
 * @author lally elias <lallyelias87@gmail.com>
 * @license MIT
 * @since  0.1.0
 * @version 0.2.0
 * @public
 */

/**
 * @apiDefine Permission
 * @apiSuccess {String} _id Unique permission identifier
 * @apiSuccess {String} resource Resource constrained by a permission
 * @apiSuccess {String} action Action(or permit) constrained(or granted)
 * by a permission
 * @apiSuccess {String} [description] A brief summary about a permission if
 * available i.e additional details that clarify what a permission for
 * @apiSuccess {String} [wildcard] System generated unique identifier of
 * a permission
 * @apiSuccess {Date} createdAt Date when permission was created
 * @apiSuccess {Date} updatedAt Date when permission was last updated
 *
 */

/**
 * @apiDefine Permissions
 * @apiSuccess {Object[]} data List of permissions
 * @apiSuccess {String} data._id Unique permission identifier
 * @apiSuccess {String} data.resource Resource constrained by a permission
 * @apiSuccess {String} data.action Action(or permit) constrained(or granted)
 * by a permission
 * @apiSuccess {String} [data.description] A brief summary about a permission if
 * available i.e additional details that clarify what a permission for
 * @apiSuccess {String} [data.wildcard] System generated unique identifier of
 * a permission
 * @apiSuccess {Date} data.createdAt Date when permission was created
 * @apiSuccess {Date} data.updatedAt Date when permission was last updated
 * @apiSuccess {Number} total Total number of permission
 * @apiSuccess {Number} size Number of permissions returned
 * @apiSuccess {Number} limit Query limit used
 * @apiSuccess {Number} skip Query skip/offset used
 * @apiSuccess {Number} page Page number
 * @apiSuccess {Number} pages Total number of pages
 * @apiSuccess {Date} lastModified Date and time at which latest permission
 * was last modified
 *
 */

/**
 * @apiDefine PermissionSuccessResponse
 * @apiSuccessExample {json} Success-Response:
 * {
 *   "_id": "5b622350480e576243f10d8b",
 *   "resource": "Task",
 *   "action": "create",
 *   "description": "In quos quae sed consectetur voluptas praesentium.",
 *   "wildcard": "task:create",
 *   "updatedAt": "2018-08-01T21:17:04.729Z",
 *   "createdAt": "2018-08-01T21:17:04.729Z"
 * }
 *
 */

/**
 * @apiDefine PermissionsSuccessResponse
 * @apiSuccessExample {json} Success-Response:
 * {
 *   "data": [
 *     {
 *       "_id": "5b622350480e576243f10d8b",
 *       "resource": "Task",
 *       "action": "create",
 *       "description": "In quos quae sed consectetur voluptas praesentium.",
 *       "wildcard": "task:create",
 *       "updatedAt": "2018-08-01T21:17:04.729Z",
 *       "createdAt": "2018-08-01T21:17:04.729Z"
 *     }
 *   ],
 *   "total": 20,
 *   "size": 10,
 *   "limit": 10,
 *   "skip": 0,
 *   "page": 1,
 *   "pages": 2,
 *   "lastModified": "2018-07-29T10:11:38.111Z"
 * }
 */

/* dependencies */
const _ = require('lodash');
const { getString } = require('@lykmapipo/env');
const {
  Router,
  getFor,
  schemaFor,
  postFor,
  getByIdFor,
  patchFor,
  putFor,
  deleteFor,
} = require('@lykmapipo/express-rest-actions');

/* constants */
const API_VERSION = getString('API_VERSION', '1.0.0');
const PATH_SINGLE = '/permissions/:id';
const PATH_LIST = '/permissions';
const PATH_SCHEMA = '/permissions/schema/';

/* declarations */
const Permission = require('./permission.model');
const router = new Router({
  version: API_VERSION,
});

/**
 * @api {get} /permissions List Permissions
 * @apiVersion 1.0.0
 * @apiName GetPermissions
 * @apiGroup Permission
 * @apiDescription Returns a list of permissions
 * @apiUse RequestHeaders
 * @apiUse Permissions
 *
 * @apiUse RequestHeadersExample
 * @apiUse PermissionsSuccessResponse
 * @apiUse JWTError
 * @apiUse JWTErrorExample
 * @apiUse AuthorizationHeaderError
 * @apiUse AuthorizationHeaderErrorExample
 */
router.get(
  PATH_LIST,
  getFor({
    get: (query, done) => {
      const options = { limit: Number.MAX_SAFE_INTEGER };
      return Permission.get(options, done);
    },
  })
);

/**
 * @api {get} /permissions/schema Get Permission Schema
 * @apiVersion 1.0.0
 * @apiName GetPermissionSchema
 * @apiGroup Permission
 * @apiDescription Returns permission json schema definition
 * @apiUse RequestHeaders
 */
router.get(
  PATH_SCHEMA,
  schemaFor({
    getSchema: (query, done) => {
      const jsonSchema = Permission.jsonSchema();
      return done(null, jsonSchema);
    },
  })
);

/**
 * @api {post} /permissions Create New Permission
 * @apiVersion 1.0.0
 * @apiName PostPermission
 * @apiGroup Permission
 * @apiDescription Create new permission
 * @apiUse RequestHeaders
 * @apiUse Permission
 *
 * @apiUse RequestHeadersExample
 * @apiUse PermissionSuccessResponse
 * @apiUse JWTError
 * @apiUse JWTErrorExample
 * @apiUse AuthorizationHeaderError
 * @apiUse AuthorizationHeaderErrorExample
 */
router.post(PATH_LIST, postFor());

/**
 * @api {get} /permissions/:id Get Existing Permission
 * @apiVersion 1.0.0
 * @apiName GetPermission
 * @apiGroup Permission
 * @apiDescription Get existing permission
 * @apiUse RequestHeaders
 * @apiUse Permission
 *
 * @apiUse RequestHeadersExample
 * @apiUse PermissionSuccessResponse
 * @apiUse JWTErrorExample
 * @apiUse AuthorizationHeaderError
 * @apiUse AuthorizationHeaderErrorExample
 */
router.get(
  PATH_SINGLE,
  getByIdFor({
    getById: (query, done) => Permission.getById(query, done),
  })
);

/**
 * @api {patch} /permissions/:id Patch Existing Permission
 * @apiVersion 1.0.0
 * @apiName PatchPermission
 * @apiGroup Permission
 * @apiDescription Patch existing permission
 * @apiUse RequestHeaders
 * @apiUse Permission
 *
 * @apiUse RequestHeadersExample
 * @apiUse PermissionSuccessResponse
 * @apiUse JWTError
 * @apiUse JWTErrorExample
 * @apiUse AuthorizationHeaderError
 * @apiUse AuthorizationHeaderErrorExample
 */
router.patch(
  PATH_SINGLE,
  patchFor({
    patch: (query, done) => {
      const options = _.pick(query, '_id', 'description');
      return Permission.patch(options, done);
    },
  })
);

/**
 * @api {put} /permissions/:id Put Existing Permission
 * @apiVersion 1.0.0
 * @apiName PutPermission
 * @apiGroup Permission
 * @apiDescription Put existing permission
 * @apiUse RequestHeaders
 * @apiUse Permission
 *
 * @apiUse RequestHeadersExample
 * @apiUse PermissionSuccessResponse
 * @apiUse JWTError
 * @apiUse JWTErrorExample
 * @apiUse AuthorizationHeaderError
 * @apiUse AuthorizationHeaderErrorExample
 */
router.put(
  PATH_SINGLE,
  putFor({
    put: (query, done) => {
      const options = _.pick(query, '_id', 'description');
      return Permission.put(options, done);
    },
  })
);

/**
 * @api {delete} /permissions/:id Delete Existing Permission
 * @apiVersion 1.0.0
 * @apiName DeletePermission
 * @apiGroup Permission
 * @apiDescription Delete existing permission
 * @apiUse RequestHeaders
 * @apiUse Permission
 *
 * @apiUse RequestHeadersExample
 * @apiUse PermissionSuccessResponse
 * @apiUse JWTError
 * @apiUse JWTErrorExample
 * @apiUse AuthorizationHeaderError
 * @apiUse AuthorizationHeaderErrorExample
 */
router.delete(PATH_SINGLE, deleteFor());

/* expose permission router */
module.exports = exports = router;
